### PR TITLE
R Chart/PDF output filenames escape '%' with '%%'

### LINKED ISSUE #1412: paths continaing odd charaters not properly escaped in call to R
### Affected tool(s)
CollectMultipleMetrics (base_distribution_by_cycle maybe others too)

### Affected version(s)
- [x] Latest public release version [2.20.6]
- [ ] Latest development/master branch as of [date of test?]

### Description 
I see this error when I try to run   
```
picard -Xmx16g CollectMultipleMetrics     PROGRAM=CollectInsertSizeMetrics     PROGRAM=CollectGcBiasMetrics     PROGRAM=CollectSequencingArtifactMetrics      I='05-1%-1.md.bam' O='05-1%-1.md.bam' R='/mnt/galaxy/data/genome/grch38_full/bwa/GRCh38_full_analysis_set_plus_decoy_hla.fa'     VALIDATION_STRINGENCY=LENIENT
```

```
INFO  2019-10-15 13:11:35     RExecutor       Executing R script via command: Rscript /tmp/langhorst/script17683804813828027089.R 05-1%-1.md.bam.base_
distribution_by_cycle_metrics 05-1%-1.md.bam.base_distribution_by_cycle.pdf 05-1%-1.md.bam
  ERROR 2019-10-15 13:11:36     ProcessExecutor Error in pdf(outputFile) :
  ERROR 2019-10-15 13:11:36     ProcessExecutor   invalid 'file' argument '05-1%-1.md.bam.base_distribution_by_cycle.pdf'
  ERROR 2019-10-15 13:11:36     ProcessExecutor Execution halted
  [Tue Oct 15 13:11:36 EDT 2019] picard.analysis.CollectMultipleMetrics done. Elapsed time: 3.45 minutes.
  Runtime.totalMemory()=1342177280
  To get help, see http://broadinstitute.github.io/picard/index.html#GettingHelp
  Exception in thread "main" picard.PicardException: R script nucleotideDistributionByCycle.R failed with return code 1
        at picard.analysis.CollectBaseDistributionByCycle.finish(CollectBaseDistributionByCycle.java:139)
        at picard.analysis.SinglePassSamProgram.makeItSo(SinglePassSamProgram.java:164)
        at picard.analysis.CollectMultipleMetrics.doWork(CollectMultipleMetrics.java:597)
        at picard.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:305)
        at picard.cmdline.PicardCommandLine.instanceMain(PicardCommandLine.java:103)
        at picard.cmdline.PicardCommandLine.main(PicardCommandLine.java:113)
```

note that running 
```
picard -Xmx16g CollectMultipleMetrics     PROGRAM=CollectInsertSizeMetrics     PROGRAM=CollectGcBiasMetrics     PROGRAM=CollectSequencingArtifactMetrics      I='05-1%-1.md.bam' O='05-1pct-1.md.bam' R='/mnt/galaxy/data/genome/grch38_full/bwa/GRCh38_full_analysis_set_plus_decoy_hla.fa'     VALIDATION_STRINGENCY=LENIENT
```
DOES work, so the problem is in the output file not the input file.

I think that path should be properly handled.

#### Steps to reproduce
name an output file with the % character

#### Expected behavior
should produce the file instead of the error.
All other tools in this pipeline handled the % character properly

#### Actual behavior
picard reports the error above

### PR DESCRIPTION
**Pull request by alanhoyle**
### Description

The output filenames for R charts now escape `%` characters to `%%`

It does this by taking running `.replaceAll("%", "%%")` on the PDF chart output filename that gets sent to the Rscript.  

This pull fixes #1412 (reported by someone else), and [my forum post](https://gatk.broadinstitute.org/hc/en-us/community/posts/360077480551-RExecutor-ProcessExecutor-invalid-file-argument-error) about the error.   

The R [`pdf()`](https://www.rdocumentation.org/packages/grDevices/versions/3.6.2/topics/pdf) function has an auto-numbering feature (`onefile=false`) that Picard does not appear to use.   However, even when not enabled, R sends the PDF output filename through a `sprintf()` function which uses the `%` character as a prefix for formatting and variable replacement.

If the filename requested by Picard contains a `%` character, this results in either Picard failing due to a `ProcessExecutor   invalid 'file' argument` error, or (rarely) malformed filenames if the filename happens to conform `sprintf()` formatting (e.g. if the filename happened to be `something-%dilution.pdf` it would result in an output file `something-1ilution.pdf`)

This issue can show up in the following Picard tools that use `RExecutor()` method

* `CollectAlignmentSummaryMetrics`
* `CollectBaseDistributionByCycle`
* `CollectGcBiasMetrics`
* `CollectInsertSizeMetrics`
* `CollectRnaSeqMetrics`
* `CollectRrbsMetrics`
* `CollectWgsMetricsWithNonZeroCoverage`
* `MeanQualityByCycle`
* `QualityScoreDistribution`

Another way to solve this would be to make the change in the RScripts in `picard/src/main/resources/analysis` by doing something like this before the `pdf(outputFile)`

```R
outputFile <- gsub('%', '%%', outputFile)
```

It also fixes a typo in a comment in the `Dockerfile` and removes trailing whitespace on a couple of lines.  

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [ ] Added or modified tests to cover changes and any new functionality
- [ ] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

**Comment by gbggrant** [comment]
Hi @alanhoyle this looks good.  Could you add a test case for this?  Perhaps add to CollectInsertSizeMetrics a test case where it's generating a pdf with an '%' in the file name?

**Review by gbggrant — CHANGES_REQUESTED** [review]
Please add a test case

**Comment by alanhoyle** [comment]
@gbggrant Please pardon my ignorance:  I'm not much of a Java developer.  What do I need to do to add a test case? 

Taking your suggestion, I assume I need to edit `CollectInsertSizeMetricsTest.java`, and I have an inkling that the `testMultipleOrientationsForHistogram()` method would be a good model to base for this test.  

How would I confirm that the test passes?  Are those things automatically run if I do a Docker build on the source tree?

**Comment by alanhoyle** [comment]
I think I've figured this out and added method `testPercentCharInPdfFilename()` to `CollectInsertSizeMetricsTest.java`.  It turns out that my dev environment (VSCode) has some pretty great default extensions that made it easy for me to run a tests.  

Screen Shot 2021-04-29 at 9 04 05 AM

File: src/test/java/picard/analysis/CollectInsertSizeMetricsTest.java [hunk 1]
```diff
@@ -264,6 +264,24 @@ public void testMultipleOrientationsForHistogram() throws IOException {
...
+
+        final String[] args = new String[] {
+            "INPUT="  + input.getAbsolutePath(),
+            "OUTPUT=" + outfile.getAbsolutePath(),
+            "Histogram_FILE=" + pdf.getAbsolutePath(),
+            "LEVEL=SAMPLE",
+            "LEVEL=LIBRARY",
+            "LEVEL=READ_GROUP"
+        };
+        Assert.assertEquals(runPicardCommandLine(args), 0);
```
Comment (gbggrant) [inline src/test/java/picard/analysis/CollectInsertSizeMetricsTest.java:282]: Would it make sense to check that the outfile is created with the expected name?  That is, still with a single %?  (just to future proof this).

**Review by gbggrant — APPROVED** [review]
Hi @alanhoyle looks good - thanks for adding the test.
It's a pity the substitution couldn't be done in a central place (like RExecutor) to get rid of duplicated code.

**Comment by gbggrant** [comment]
Oh, and the 'Branch' build is currently not working on forks, so if you're all set with this (that is you don't have anything to add), just say so and I'll do the merge to master.

File: src/test/java/picard/analysis/CollectInsertSizeMetricsTest.java (hunk 1 above)
Comment (alanhoyle) [reply src/test/java/picard/analysis/CollectInsertSizeMetricsTest.java:282]: It would.  I'll add that in a new commit.

Comment (alanhoyle) [reply src/test/java/picard/analysis/CollectInsertSizeMetricsTest.java:282]: Added below.

**Review by alanhoyle — COMMENTED** [review]
I've added an additional test immediately below the selected lines.

**Comment by alanhoyle** [comment]
> Hi @alanhoyle looks good - thanks for adding the test.
> It's a pity the substitution couldn't be done in a central place (like RExecutor) to get rid of duplicated code.

I don't think you can put it in `RExecutor()` since the parameters passed to the R scripts vary from tool to tool.  If you do a blanket substitution in there, you'd end up breaking things where it wouldn't know how to find the input files used to generate the figures.  

As I said in the original request, it might make slightly more sense to incorporate this directly into the R scripts in the `pdf(outfile)` calls, but that doesn't solve the duplicated code issue.  I note that in one of the R scripts (`gcBias.R`), it explictly says `pdf(outputFile, onefile=TRUE);`  and the `onefile` feature implementation seems to be the thing that causes the problematic behavior.  If RProject changes the behavior to where when `onefile=TRUE` that it doesn't do the `sprintf()` on the filename, it would result in Picard doubling the `%` chars in the PDF filenames, but it would at least still run to completion.

**Comment by gbggrant** [comment]
> > Hi @alanhoyle looks good - thanks for adding the test.
> > It's a pity the substitution couldn't be done in a central place (like RExecutor) to get rid of duplicated code.
> 
> I don't think you can put it in `RExecutor()` since the parameters passed to the R scripts vary from tool to tool. If you do a blanket substitution in there, you'd end up breaking things where it wouldn't know how to find the input files used to generate the figures.
> 
> As I said in the original request, it might make slightly more sense to incorporate this directly into the R scripts in the `pdf(outfile)` calls, but that doesn't solve the duplicated code issue. I note that in one of the R scripts (`gcBias.R`), it explictly says `pdf(outputFile, onefile=TRUE);` and the `onefile` feature implementation seems to be the thing that causes the problematic behavior. If RProject changes the behavior to where when `onefile=TRUE` that it doesn't do the `sprintf()` on the filename, it would result in Picard doubling the `%` chars in the PDF filenames, but it would at least still run to completion.

Yeah, I realized It couldn't go in RExecutor.  Thanks for the good work.

**Comment by alanhoyle** [comment]
Before you merge this, I also have a few changes to the `.gitignore` that others might find useful if they use a similar build environment.  (see below)

Should I commit/add those, do a separate pull request, or just move those into my global `.gitignore`

```
*.class
.project
org.eclipse*
bin/
.classpath
*.swp
```

**Comment by gbggrant** [comment]
I think we'd prefer that you not commit your '.gitignore' - those all seem fairly specific to your environment and so might cause others confusion.  At the minimum it should go in another PR.

**Pull request merged**